### PR TITLE
steps for running experiments on amazon review

### PR DESCRIPTION
**Pull request by GentleZhu**
*Issue #, if available:*

*Description of changes:*
Add a folder for reproducing graphstorm paper experiments.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/paper_experiments/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+```
+python construct_AR_items.py
+python construct_AR_reviews.py
+```
+
+### Run gconstruct on process_data
+In the first step, the processed data are under processed_data/amazon_review/
+Once the data are processed, run the following command to construct a graph
+for node classification and link prediction tasks. 
+Recommended machine: x1e.32xlarge
```
Comment (jalencato) [inline examples/paper_experiments/README.md:24]: We may need to add a link for x1e.32xlarge here for reference if anyone wants to refer the instance type for x1e.32xlarge. https://aws.amazon.com/ec2/instance-types/x1e/

Comment (classicsong) [reply examples/paper_experiments/README.md:24]: Can you profile the maximum memory requirement to process the data?

File: examples/paper_experiments/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+python construct_AR_reviews.py
+```
+
+### Run gconstruct on process_data
+In the first step, the processed data are under processed_data/amazon_review/
+Once the data are processed, run the following command to construct a graph
+for node classification and link prediction tasks. 
+Recommended machine: x1e.32xlarge	
+
+The command takes `AR_Video_Games.json` that specifies the input data for graph
```
Comment (classicsong) [inline examples/paper_experiments/README.md:26]: The AR_Video_Games.josn file is missing.

File: examples/paper_experiments/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+construction, constructs the graph, and saves the parition to `amazon_review`.
+
+```
+python -m graphstorm.gconstruct.construct_graph \
+           --conf-file gconstruct_configs/AR_construction_full.json \
+           --output-dir datasets/amazon_review_full --graph-name amazon_review \
+           --num-processes 16 --num-parts 4  --logging-level debug \
+           --skip-nonexist-edges --add-reverse-edges
+```
+The processed data is stored under datasets/amazon_review_full/
```
Comment (classicsong) [inline examples/paper_experiments/README.md:36]:
```suggestion
The processed data is stored under datasets/amazon_review_full/. The script partitions the graph into four parts.
```

File: examples/paper_experiments/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+### Run graphstorm training 
+By default, we use four instances for training and their ips are written in ip_list.txt
```
Comment (classicsong) [inline examples/paper_experiments/README.md:39]: You can provide a link to a doc about how to set up distributed training.

File: examples/paper_experiments/construct_AR_items.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    p_text = ' '.join(p_text.split('\n'))
+    p_text = ' '.join(p_text.split('\t'))
+    p_text = ' '.join(p_text.split('\rm'))
+    p_text = ' '.join(p_text.split('\r'))
+    p_text = ''.join(p_text.split('$'))
+    p_text = ''.join(p_text.split('*'))
+
+    return p_text
+
+def construct_one_domain(full_file_path, asin_dict, asin_data, edge_dict, dup_asin, brand_dict, category_dict_list, item_brand, item_category):
```
Comment (classicsong) [inline examples/paper_experiments/construct_AR_items.py:32]: Can you add some comments about what you are doing in this func?

File: examples/paper_experiments/construct_AR_reviews.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    p_text = ' '.join(p_text.split('\n'))
+    p_text = ' '.join(p_text.split('\t'))
+    p_text = ' '.join(p_text.split('\rm'))
+    p_text = ' '.join(p_text.split('\r'))
+    p_text = ''.join(p_text.split('$'))
+    p_text = ''.join(p_text.split('*'))
+
+    return p_text
+
+def construct_one_domain(full_file_path, asin_dict, asin_data, edge_dict, dup_asin, brand_dict, category_dict_list, item_brand, item_category):
```
Comment (classicsong) [inline examples/paper_experiments/construct_AR_reviews.py:31]: Ditto.

**Review by classicsong — COMMENTED** [review]
In general, we need to have a README describing the purpose of each yaml file under training_configs and the performance numbers.

File: examples/paper_experiments/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
+This folder contains the data processing and code to reproduce the graphstorm paper results.
```
Comment (classicsong) [inline examples/paper_experiments/README.md:1]: As the paper is not published. We should say:
This folder contains the data processing and code to profile GraphStorm framework on Amazon Review dataset.

File: examples/paper_experiments/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+### Process raw dataset into parquet files
+We first download the full Amazon Review dataset from https://cseweb.ucsd.edu/~jmcauley/datasets/amazon_v2/. 
+Review data is in https://datarepo.eng.ucsd.edu/mcauley_group/data/amazon_v2/categoryFiles/All_Amazon_Review.json.gz
+Meta data is in https://datarepo.eng.ucsd.edu/mcauley_group/data/amazon_v2/metaFiles/All_Amazon_Meta.json.gz
+```
+wget https://datarepo.eng.ucsd.edu/mcauley_group/data/amazon_v2/metaFiles2/meta_Video_Games.json.gz \
+-P raw_data/
+```
+
+After you unzip the above two files, we generate the parquet files as input of graph construction step.
```
Comment (classicsong) [inline examples/paper_experiments/README.md:14]:
```suggestion
After you unzip the above two files, run the following two scripts to generate the input for the graph construction step.
```

File: examples/paper_experiments/README.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+After you unzip the above two files, we generate the parquet files as input of graph construction step. 
+```
+python construct_AR_items.py
+python construct_AR_reviews.py
+```
+
+### Run gconstruct on process_data
+In the first step, the processed data are under processed_data/amazon_review/
```
Comment (classicsong) [inline examples/paper_experiments/README.md:21]:
```suggestion
After the first step, the processed data are under processed_data/amazon_review/
```

File: examples/paper_experiments/construct_AR_items.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    """ 
+    Process item-related edges and node labels. Filter out low-frequency product types and brand.
```
Comment (classicsong) [inline examples/paper_experiments/construct_AR_items.py:34]: Can you be more specific about what does low-frequency mean? 
For example if an entity appears less than 5 times, then it is treated as low-frequency.

File: examples/paper_experiments/construct_AR_items.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            if cate[1] in category_dict_list[1]:
+                pt_lvl2.append(category_dict_list[1][cate[1]])
+            else:
+                pt_lvl2.append(math.nan)
+            if cate[2] in category_dict_list[2]:
+                pt_lvl3.append(category_dict_list[2][cate[2]])
+            else:
+                pt_lvl3.append(math.nan)
+        else:
+            pt_lvl1.append(math.nan)
```
Comment (classicsong) [inline examples/paper_experiments/construct_AR_items.py:112]: Why we append math.nan here?

Comment (GentleZhu) [reply examples/paper_experiments/construct_AR_items.py:112]: NAN is required by gsprocessing to ignore these labels in train/val/test split

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**